bitarray/_bitarray.c

```diff
@@ -87,13 +94,26 @@ static PyTypeObject Bitarraytype;
 
 #define BYTES(bits)  (((bits) == 0) ? 0 : (((bits) - 1) / 8 + 1))
 
-#define BITMASK(endian, i)  (((char) 1) << ((endian) ? (7 - (i)%8) : (i)%8))
+#define BITMASK(endian, i)  (((unsigned char) 1) << ((endian) ? (7 - (i)%8) : (i)%8))
 
 /* ------------ low level access to bits in bitarrayobject ------------- */
 
 #define GETBIT(self, i)  \
     ((self)->ob_item[(i) / 8] & BITMASK((self)->endian, i) ? 1 : 0)
 
+#if HAS_VECTORS
+/*
+ * Perform bitwise operation OP on 16 bytes of memory at a time.
+ */
+#define vector_op(A, B, OP) do {  \
+    vec __a, __b, __r;            \
+    memcpy(&__a, A, sizeof(vec)); \
```

Comment:
Ah, no I haven't done anything to guarantee alignment. I did not that during the initial patch we perform an aligned move (`movdqa`) which is largely coincidental. Now it seems I largely get `movqdu`, which is the unaligned version. It doesn't appear to affect performance much with my limited test case.

Looking at the assembly loop it largely seems to do a pretty good job at the moment.
There is one odd move which I would have expected it to optimize out though that I noticed when looking closer:

```asm
.L784:
        movq    %rdx, %rax
.L790:
        leaq    -16(%rax), %rcx
        movq    32(%rbx), %rdi ; (SELF)->ob_item
        movq    %rcx, %rdx     ; (OTHER)->ob_item
        addq    32(%rbp), %rdx
        movdqu  -16(%rdi,%rax), %xmm0
        movdqu  (%rdx), %xmm1
        pand    %xmm1, %xmm0
        movups  %xmm0, (%rdx)
        leaq    16(%rax), %rdx
        cmpq    %rdx, %rsi
        ja      .L784
```

After the latest patch (0b09add), this becomes:

```asm
.L796:
        movq    %rcx, %rax
.L822:
        movdqu  -16(%rdx,%rax), %xmm0
        leaq    16(%rax), %rcx
        movdqu  -16(%rsi,%rax), %xmm1
        cmpq    %rcx, %rdi
        pand    %xmm1, %xmm0
        movups  %xmm0, -16(%rdx,%rax)
        ja      .L796
```

Lo and behold, I go from:
```
0.521735191345
0.509098052979
0.512199163437
```
To:
```
0.393866062164
0.399346828461
0.393866062164
```